sphinx/ext/autodoc/__init__.py

```diff
@@ -1056,7 +1056,11 @@ def format_args(self, **kwargs: Any) -> str:
             self.env.app.emit('autodoc-before-process-signature', unwrapped, False)
             sig = inspect.signature(unwrapped)
             args = stringify_signature(sig, **kwargs)
-        except TypeError:
+        except TypeError as exc:
+            logger.warning(__("Failed to get a function signature of %s: %s"),
+                           self.fullname, exc)
```

Comment:
Perhaps better as:
```suggestion
        except TypeError:
            logger.warning(__("Failed to get a function signature of %s"),
                           self.fullname, exc_info=True)
```

Reply:
Thank you for comment. I changed the word "for".

On the other hand, I got the following warning with `exc_info=True`:
```
WARNING: Failed to get a function signature for errno.EINTR
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1057, in format_args
    sig = inspect.signature(unwrapped)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 417, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2216, in _signature_from_callable
    raise TypeError('{!r} is not a callable object'.format(obj))
TypeError: 4 is not a callable object
```

I don't understand the worth of stacktrace of `inspect.signature()`. Why do you want to show stacktrace here?

Reply:
Because the failure may not be from `inspect.signature` at all, but from `autodoc-before-process-signature` which could throw TypeError due to user mistakes.

Reply:
Sphinx does not expect event handlers to exit with exception. If so, it's a bug of extensions. To handle it correctly, we should also handle other exceptions like following:
```
try:
     self.env.app.emit('autodoc-before-process-signature', unwrapped, False)
except Exception:
    ...

try:
    sig = inspect.signature(unwrapped)
    args = stringify_signature(sig, **kwargs)
except TypeError:
   ...
except ValueError:
   ...
```

Reply:
I made https://github.com/sphinx-doc/sphinx/pull/7658 to handle errors on event handlers.